crates/iota-indexer/src/store/pg_indexer_analytical_store.rs

```diff
@@ -480,29 +480,40 @@ impl IndexerAnalyticalStore for PgIndexerAnalyticalStore {
 
 fn construct_checkpoint_tx_count_query(start_checkpoint: i64, end_checkpoint: i64) -> String {
     format!(
-        "With filtered_txns AS (
+        "WITH expanded_checkpoint_range AS (
+            SELECT
+                sequence_number AS checkpoint_sequence_number,
+                epoch,
+                generate_series(min_tx_sequence_number, max_tx_sequence_number) AS tx_sequence_number
+            FROM checkpoints
+            WHERE sequence_number >= {start_checkpoint} AND sequence_number < {end_checkpoint}
+        ), filtered_txns AS (
             SELECT
                 t.checkpoint_sequence_number,
-                c.epoch,
+                (SELECT ecr.epoch FROM expanded_checkpoint_range ecr
+                 WHERE ecr.tx_sequence_number = t.tx_sequence_number
+                 LIMIT 1) AS epoch,
                 t.timestamp_ms,
                 t.success_command_count
             FROM transactions t
-            LEFT JOIN checkpoints c
-            ON t.checkpoint_sequence_number = c.sequence_number
-            WHERE t.checkpoint_sequence_number >= {} AND t.checkpoint_sequence_number < {}
-          )
-          INSERT INTO tx_count_metrics
-          SELECT
+            WHERE EXISTS (
+                SELECT 1
+                FROM expanded_checkpoint_range ecr
+                WHERE t.tx_sequence_number = ecr.tx_sequence_number
+            )
+        )
+        INSERT INTO tx_count_metrics
+        SELECT
             checkpoint_sequence_number,
             epoch,
             MAX(timestamp_ms) AS timestamp_ms,
             COUNT(*) AS total_transaction_blocks,
             SUM(CASE WHEN success_command_count > 0 THEN 1 ELSE 0 END) AS total_successful_transaction_blocks,
             SUM(success_command_count) AS total_successful_transactions
-          FROM filtered_txns
-          GROUP BY checkpoint_sequence_number, epoch ORDER BY checkpoint_sequence_number
-          ON CONFLICT (checkpoint_sequence_number) DO NOTHING;
-        ", start_checkpoint, end_checkpoint
+        FROM filtered_txns
+        GROUP BY checkpoint_sequence_number, epoch
+        ORDER BY checkpoint_sequence_number
+        ON CONFLICT (checkpoint_sequence_number) DO NOTHING;"
```

Comment:
Wouldn't it be simpler and likely more efficient to simplify to
```sql
)

INSERT INTO tx_count_metrics
SELECT
    ecr.checkpoint_sequence_number,
    ecr.epoch,
    MAX(t.timestamp_ms) AS timestamp_ms,
    COUNT(*) AS total_transaction_blocks,
    SUM(CASE WHEN t.success_command_count > 0 THEN 1 ELSE 0 END) AS total_successful_transaction_blocks,
    SUM(t.success_command_count) AS total_successful_transactions
FROM expanded_checkpoint_range ecr
JOIN transactions t
    ON t.tx_sequence_number = ecr.tx_sequence_number
GROUP BY ecr.checkpoint_sequence_number, ecr.epoch
ORDER BY ecr.checkpoint_sequence_number
ON CONFLICT (checkpoint_sequence_number) DO NOTHING;

```
That is remove the `filtered_txns` and `JOIN` directly on `transactions`?

Reply:
You're absolutely right @kodemartin !

Reply:
Fixed with https://github.com/iotaledger/iota/pull/5386/commits/a215e77d6945679177a79d94981d20e4930a6717